observers/qos_correction.cpp

```diff
@@ -0,0 +1,200 @@
+              return false;
+
+            aggressorsToKill.push_back(
+                createKill(possibleAggressor.executor_info()));
+
+            return true;
+          });
+      continue;
+
+    } else {
+      severity += severity_balance;
+      if (severity > 0) {
+        // Assuming that severity is reflecting how many
+        // resources is needed for PR job to not starve again.
+        // TODO(bplotka) We could implement here more sophisticated algorithm
+        // e.g: solving discrete knapsack problem.
+        // Currently we are killing most active BE tasks first.
+        // (Greedy behaviour).
+        possibleAggressors.remove_if(
+            [&severity, &aggressorsToKill]
+                (ResourceUsage_Executor& possibleAggressor) {
+              if (severity <= 0)
+                return false;
+
+              Resources aggressorsAllocation(possibleAggressor.allocated());
+              if (aggressorsAllocation.cpus().isSome()) {
+                severity -= aggressorsAllocation.cpus().get();
+
+                LOG(INFO) << "Decided to kill Executor: " <<
+                possibleAggressor.executor_info().executor_id();
+
+                aggressorsToKill.push_back(
+                    createKill(possibleAggressor.executor_info()));
+              }
+
+              return true;
+            });
+      }
+
+      if (severity > 0) {
+        // In such case even if we kill all BE executors contention is not
+        // solved.
+        LOG(ERROR) << "Aggressors are not the cause of CPU contention"
+            " or lack of info about some aggressors.";
+        break;
+      }
+
+      // We often kill executor with usage > severity. Keep this balance
+      // for another contentions.
+      severity_balance += severity;
+    }
+    // Eliminating duplicate contentions.
+    correctedPrExecutors.push_back(WID(contention.victim()));
+  }
+
+  // Create QoSCorrection having aggressors list.
+  for (auto aggressorToKill : aggressorsToKill) {
+    corrections.push_back(createKillQoSCorrection(aggressorToKill));
+  }
+
+  return corrections;
+}
+
+
+QoSCorrectionObserver::~QoSCorrectionObserver() {}
+
+
+Try<Nothing> QoSCorrectionObserver::_consume(
+  const std::vector<Contentions> products) {
+  Contentions newContentions;
+
+  for (Contentions contentions : products)
+    for (Contention contention : contentions)
+      newContentions.push_back(contention);
+
+  this->currentContentions = newContentions;
+  this->currentContentions.get().sort(compareSeverity);
+
+  if (this->currentUsage.isSome()) {
+    this->__correctSlave();
+  }
+
+  return Nothing();
+}
+
+
+Try<Nothing> QoSCorrectionObserver::consume(const ResourceUsage& usage) {
+  this->currentUsage = usage;
+
+  if (this->currentContentions.isSome()) {
+    this->__correctSlave();
+  }
+
+  return Nothing();
+}
+
+
+Try<Nothing> QoSCorrectionObserver::__correctSlave() {
+  CHECK_ERR(this->currentContentions.isSome());
```

Comment:
Nope - it will bring down the QoSController - not whole slave. I tested it. But you're right that it should be a soft error (: